### PR TITLE
[BEAM-10706] Fix duplicate key error in DynamoDBIO.Write

### PR DESCRIPTION
**Pull request by dennisylyung**
DynamoDBIO.Write will throw a duplicate key error when write requests in the batch contains duplicate keys. 
An optional deduplication mechanism is added to avoid this error, taking reference from the implementation in [the official Python SDK of DynamoDB](https://github.com/boto/boto3/blob/develop/boto3/dynamodb/table.py).
Changes to the ver1 module (beam-sdks-java-io-amazon-web-services) has been tested on an actual DynamoDB table with writes with duplicate keys. The job throws duplicate error if overwriteByPKeys are not specified, but ran successfully if the table's partition and sort keys are provided, as expected. 
Changes have been ported to the ver2 module (beam-sdks-java-io-amazon-web-services2), but not tested on an actual dynamodb table. It passed local unit tests as did v1. 
tests are updated to include provision of overwriteByPKeys, but a dedicated test for duplicate keys has not been written, since the mock dynamodb does not seem to throw on duplicate keys anyhow. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by dennisylyung** [comment]
R:   @jbonofre

**Comment by iemejia** [comment]
Hi thanks for your contribution(s), sorry if I have been slow with this (and the other) reviews.

I have a minor question here. Is this because a limitation of AWS API level at the batched request level only? Or is deduplication required at the whole table level because if it is like that this fix won't make it, but including a filter like this as a transform would make sense.

**Comment by dennisylyung** [comment]
@iemejia The limitation is on the API level, but not the table level. AWS support has confirmed this to me, and suggest performing deduplication of the batches before sending the request with the Java SDK. The Python SDK for DynamoDB does this on the SDK level (before API), but not the Java one, so I just cloned the deduplication mechanism.

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java [hunk 1]
```diff
@@ -342,6 +343,8 @@ public boolean test(Throwable throwable) {
 
     abstract @Nullable SerializableFunction<T, KV<String, WriteRequest>> getWriteItemMapperFn();
 
+    abstract @Nullable List<String> getOverwriteByPKeys();
```
Comment (iemejia) [inline sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:346]: Can you please remove the `@Nullable` here and set an empty collection as the default value on the `build()` method. This will simplify the validations in the rest of the code and will make it simpler.

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java [hunk 2]
```diff
@@ -450,12 +459,35 @@ public void startBundle(StartBundleContext context) {
...
+        if (spec.getOverwriteByPKeys() != null) {
+          removeDupPKeysRequestsIfAny(writeRequest.getValue());
+        }
         batch.add(writeRequest);
         if (batch.size() >= BATCH_SIZE) {
           flushBatch();
         }
       }
 
+      private void removeDupPKeysRequestsIfAny(WriteRequest request) {
```
Comment (iemejia) [inline sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:471]: Can we better restructure this method to instead of removing from the batch every time we just skip adding the value if it already is part of the batch. It could make the code simpler.

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java [hunk 3]
```diff
@@ -450,12 +459,35 @@ public void startBundle(StartBundleContext context) {
...
+        Map<String, AttributeValue> pKeyValueNew = extractPkeyValues(request);
+        batch.removeIf(item -> extractPkeyValues(item.getValue()).equals(pKeyValueNew));
```
Comment (iemejia) [inline sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:473]: I am wondering if this will get the existing key removed too? Maybe worth to validate this in the tests.

File: sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java [hunk 4]
```diff
@@ -186,6 +192,10 @@ public void testRetries() throws Throwable {
 
     final List<WriteRequest> writeRequests = DynamoDBIOTestHelper.generateWriteRequests(numOfItems);
 
+    final List<String> overwriteByPKeys = new ArrayList<>();
```
Comment (iemejia) [inline sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java:195]: nit: `Arrays.asList` will make the code more concise.

File: sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java [hunk 5]
```diff
@@ -199,7 +209,8 @@ public void testRetries() throws Throwable {
                         writeRequest -> KV.of(tableName, writeRequest))
                 .withRetryConfiguration(
                     DynamoDBIO.RetryConfiguration.create(4, Duration.standardSeconds(10)))
-                .withAwsClientsProvider(AwsClientsProviderMock.of(amazonDynamoDBMock)));
+                .withAwsClientsProvider(AwsClientsProviderMock.of(amazonDynamoDBMock))
+                .withOverwriteByPKeys(overwriteByPKeys));
```
Comment (iemejia) [inline sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java:213]: Can we do this in an independent test to not change the original write test, e.g. `testDeduplicateWriteItems` and produce some duplicate items that test the correct removal of keys behaviour. I am assuming that for the batch case the direct runner will batch them together, but I will let a further note when I confirm this 100%.

**Review by iemejia — CHANGES_REQUESTED** [review]
Thanks for bringing this fix. I let some suggestions to try to make the implementation and the tests simpler. I let comments only in the AWS impl but they apply too for the v2 version.

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java (hunk 2 above)
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:471]: I think the later event is more preferred, although Apache beam do not have strict ordering. Not sure why the AWS Python SDK is implemented this way. I will try using something like a map.

File: sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java (hunk 5 above)
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java:213]: Spent some time working on this but couldn't come up with anything. Any tips how to test if the batches are correct? I can't think of anyway to test it in a pipeline, as the end result written to the db is the same (the issue is on api level). Should I do it outside a pipeline, call processElement manually and check the batch size?

**Comment by iemejia** [comment]
I have forgotten about this issue, did you have any progress @dennisylyung, any chance you can get this rebased so we can have some progress.
I also just realized that you made a question that I have not answered too :S I have to check this on my side and tell you back.

**Comment by dennisylyung** [comment]
@iemejia I have fixed most of your previous comments locally, but not the last one about the test. 
I am not too familiar with the Java test suite. (I just know that in Python I can do something like MagicMock).
If you can point the directions, I think I have some capacity to work on this in the coming days though.

**Comment by iemejia** [comment]
Oh maybe I was not clear, I did not expect anything fancier, only two things: (1) that we create a specific test for the duplicateKey case and not do it in the generic write test (you can basically reuse the code but in a different method not to affect the normal path, and (2) that you pass duplicate keys as part of the write to see if deduplication is working. That should be enough. The sole doubt I have is how can we guarantee that the keys are part of the same batch. I assume that `DirectRunner` will put them by default in the same bundle. I suppose we can debug to verify this. In the past we used a class called `DoFnTester` for cases like this but I don't know of a more modern alternative to do it.

**Comment by dennisylyung** [comment]
@iemejia I made some changes to use a hashmap for deduplication, and to improve the test. 
I separated the test for duplicate key, and used mockito to check whether the call to the aws client contains duplicate key. 
However, as you have worried, each element is put into a single bundle. Hence each bundle consist of one element only, defeating the duplication test.
I looked at the documentation of `DoFnTester` but it is deprecated. The docs suggests using `TestPipeline` instead, but I couldn't find anyway to control the bundling of elements. 
Any idea?

**Comment by dennisylyung** [comment]
@iemejia I have wrote a separate test for the deduplication, using Mockito to verify the contents submitted to the AWS client. Please check.

**Comment by dennisylyung** [comment]
@iemejia The PR should be ready, please review.

**Comment by iemejia** [comment]
@dennisylyung thanks sorry I have been busy with other stuff but I will take a look.

**Comment by dennisylyung** [comment]
@iemejia I have some some larger projects coming up end of the year. Would you be able to give feedbacks soon, such that I can work on follow up changes before I then? Thanks!
Since we are relying on this patch, I hope that it can be merged so I won't need to maintain my patched build.

**Comment by iemejia** [comment]
My excuses @dennisylyung I took a quick look today and it looks mostly ok (I have some minor things that I can fix) the only issue I have is that I just realized that the tests are disabled for DynamoDB how did you get them to run? I removed the Ignore annotation and enabled the specific extra test on this PR but I get a failure.

**Comment by dennisylyung** [comment]
> My excuses @dennisylyung I took a quick look today and it looks mostly ok (I have some minor things that I can fix) the only issue I have is that I just realized that the tests are disabled for DynamoDB how did you get them to run? I removed the Ignore annotation and enabled the specific extra test on this PR but I get a failure.

I tested also by commenting out the ignore annotation.
The test itself passes, but `destroy` fails due to this bug https://github.com/localstack/localstack/issues/2702. The error is universal to all tests in DynamoDBIOTest, not just the new one.

I upgraded Localstack to the fixed version and it now passes. (It always passes in aws2) @iemejia

**Comment by iemejia** [comment]
Thanks a lot for the info on localstack if you feel motivated please open a PR to update localstack and enable the tests that work in that PR otherwise I will do.

I am trying to understand this again because it seems I may have misinterpreted it before. If we need deduplication to guarantee that operations do not fail in the presence of duplicate attributes why we need to pass the overwrite keys explictly? With the current implementation I have the impression we could end up filtering also non repeated keys. Couldn't we just simply compare the current `private List<KV<String, WriteRequest>> batch;` with the current element  and overwrite the duplicate element if is the case. I think this will make the implementation much simpler (and also won't change order).

Also if deduplication is mandatory to not have `ValidationException` shouldn't this be the default behaviour? Or in which case I can benefit of avoiding deduplication?

I excuse myself because I don't want to delay this longer but I still think I do not understand the fix.

**Comment by dennisylyung** [comment]
In the current implementation `private List<KV<String, WriteRequest>> batch`, the key is the table name, not the primary-key. 

for example, in a table `user`, the primary key is `id`. A batch entry would be like this: 
`KV("user", {id=1, name=Chris, age=30})`
We have no way to know that `id` is the key we need to deduplicate on without users specifying. 

Theoretically, operating with a DynamoDB should not require setting the keys for de-duplication, since repeated write to the same key will just update the value. However, the current implementation of the DynamoDB batch put API requires no duplicate keys within a batch. Hence, users need to explicitly set the overwrite keys. 

You are right that the overwrite keys are necessary to completely avoid `ValidationError`. As long as the sink operate in upsert logic (i.e. the data could contain duplicate keys), there is a risk of the same keys going into a single batch. This is also the problem I face developing pipelines with DynamoDB sinks.

There is one special case though. If the user is very sure that the keys will never have duplicates, such as when their pipelines are logically append-only, they will not encounter `ValidationError`. In which case, requiring them to specify the keys could be unnecessary.

**Comment by dennisylyung** [comment]
I have opened another PR for the localstack upgrade #13510

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by iemejia** [comment]
@dennisylyung are you on the ASF slack? I would like to ask you some questions in real time if possible.

**Review by iemejia — CHANGES_REQUESTED** [review]
Thanks for the explanation! It is probably a good idea to add this to the Javadoc. I let some comments in the PR review, please apply those too to the AWSv2 version.

I still have a doubt about the composed keys question I let below.

**Comment by iemejia** [comment]
Beam next release will be cut in 5 days, so I think we still have time to get this in, again sorry for the delay.

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java [hunk 6]
```diff
@@ -447,19 +467,41 @@ public void setup() {
...
         if (batch.size() >= BATCH_SIZE) {
           flushBatch();
         }
       }
 
+      private Map<String, AttributeValue> extractPkeyValues(WriteRequest request) {
+        if (spec.getOverwriteByPKeys() != null) {
+          if (request.getPutRequest() != null) {
+            return request.getPutRequest().getItem().entrySet().stream()
+                .filter(entry -> spec.getOverwriteByPKeys().contains(entry.getKey()))
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:488]: This should handle composite keys. 

Supposedly, this gets the item map
```
request.getPutRequest().getItem()        // Map<String, AttributeValue>
```
and then retains only the relevant keys 
```
... .entrySet().stream()
    .filter(...)
    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
```
The filter function checks whether the key is one of the specified deduplication keys
```
entry -> spec.getOverwriteByPKeys().contains(entry.getKey())
```

Is this correct?

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java [hunk 7]
```diff
@@ -85,7 +87,8 @@
  *                       t -> KV.of(tableName, writeRequest))
  *               .withRetryConfiguration(
  *                    DynamoDBIO.RetryConfiguration.create(5, Duration.standardMinutes(1)))
- *               .withAwsClientsProvider(new BasicDynamoDbProvider(accessKey, secretKey, region));
+ *               .withAwsClientsProvider(new BasicDynamoDbProvider(accessKey, secretKey, region))
+ *               .withOverwriteByPKeys(overwriteByPKeys));
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:91]: Definitely, overwriteByPKeys isn't too intuitive

[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java:91]: Sure

**Review by iemejia — APPROVED** [review]
LGTM Great job @dennisylyung!

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (iemejia) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java]: Yes it seems I might have misinterpreted, clear now.

**Comment by iemejia** [comment]
Merged manually to squash the commits. Thanks again @dennisylyung and don't hesitate to contribute more in the future I will be more reactive or pass the review to someone with more time.

**Comment by dennisylyung** [comment]
Thanks for the review! @iemejia
